core-rust/state-manager/src/protocol/protocol_config.rs

```diff
@@ -0,0 +1,161 @@
+use radix_engine::prelude::{ScryptoCategorize, ScryptoDecode, ScryptoEncode};
+use radix_engine_common::math::Decimal;
+use radix_engine_common::prelude::{hash, scrypto_encode};
+
+use radix_engine_common::types::Epoch;
+
+use crate::StateVersion;
+use utils::btreeset;
+
+// This file contains types for node's local static protocol configuration
+
+const MAX_PROTOCOL_VERSION_NAME_LEN: usize = 15;
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub struct ProtocolUpdate {
+    pub next_protocol_version: String,
+    pub enactment_condition: ProtocolUpdateEnactmentCondition,
+}
+
+impl ProtocolUpdate {
+    pub fn readiness_signal_name(&self) -> String {
+        // Readiness signal name is 32 bytes long and consists of:
+        // - protocol version name (up to 15 bytes)
+        // - a one byte separator ("-")
+        // - 16+ bytes of `hash(enactment_condition + next_protocol_version)`
+        let mut bytes_to_hash = scrypto_encode(&self.enactment_condition).unwrap();
+        bytes_to_hash.extend_from_slice(self.next_protocol_version.as_bytes());
+        let protocol_update_hash = hash(&bytes_to_hash);
+        let mut res = self.next_protocol_version.to_owned();
+        res.push('-');
+        res.push_str(&hex::encode(protocol_update_hash)[0..(32 - res.len())]);
+        res
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub struct ProtocolConfig {
+    pub genesis_protocol_version: String,
+    pub protocol_updates: Vec<ProtocolUpdate>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub enum ProtocolUpdateEnactmentCondition {
+    EnactWhenSupportedAndWithinBounds {
+        lower_bound: ProtocolUpdateEnactmentBound,
+        upper_bound: ProtocolUpdateEnactmentBound,
+        support_type: ProtocolUpdateSupportType,
+    },
+    EnactUnconditionallyAtEpoch(Epoch),
+    EnactUnconditionallyAtStateVersion(StateVersion),
```

Comment:
just as discussed on our call: it still seems to me like some portion of conceptual complexity (and code) could be gone if we discarded the "unconditional" methods (and used `required_ratio_of_stake_supported: 0.0` for "asap updates"; and stored the "already-enacted-by-quorum" updates in a map). Please see if there are any true blockers to migrate to that approach.